- Some callback script added, changes in active_standby.py to shutdown Standby if it comes up unexpectedly

cornec81 opened the pull request:
- Added email_callback.py for email functionality
- Added client1 example files for sourcing env vars, callbacks and launcher/runner
- Changed ative_standby.py to subscribe to events on Standby PC and shut it down if someone tries to unshut it
- some other typos and corrections

Commit:
- Added email_callback.py for email functionality
- Added client1 example files for sourcing env vars, callbacks and launcher/runner
- Changed ative_standby.py to subscribe to events on Standby PC and shut it down if someone tries to unshut it
- some other typos and corrections

Commit:
- updated client1_runner.sh to only run single instance

amney commented:
Hey Corne - great pull request, glad you guys are really running with it.

Can we discuss one piece of logic you have added? Subscribing to the standby pc and then blacklisting it if you see it come up

I'm concerned that a small (but possible) combination with the wrong set of timing and events you could end up in a sticky situation, specifically when you run this in HA:
- two scripts monitoring active / standby
- active pc fails
- script one detects it first and un-shuts standby
- script one exits
- network / apic issue causes script two to not receive the active down event
- script two receives the standby up event and disables the standby link
- now both links are shut 

It's pretty (very) unlikely but something we should consider? Personally I think it's worth the risk because you are more likely to face someone accidentally bringing up the standby link

amney commented on active_standby.py:
```diff
@@ -45,70 +45,107 @@ def refresh_login():
+        
+            if sub_id == active_standby.active_sub_id:
+                # the event was on the subscription for the active link
+                try:
```
Can we pull this `try:` statement a bit higher up to line 48. This should help catch any errors Key or Value errors gracefully within this entire `on_message` function.

amney commented on client1.env:
```diff
@@ -0,0 +1,13 @@
+SBY="topology/pod-1/node-106/sys/aggr-[po6]"
+
+# callback must be absolute path
+CALLBACK=/home/admin/active_standby/client1_callback.sh
```
I didn't highlight it well in the README document, but the script actually has built in support for env variables (so you do not need to pass them as cli parameters)

Basically, prefix all env variables with AS and then slugify the cli parameter

For example

```
AS_APIC_ADDRESS=127.0.0.1  
AS_APIC_USER=admin 
AS_PC_ACTIVE="topology/pod-1/node-105/sys/aggr-[po9]"
AS_PC_STANDBY="topology/pod-1/node-106/sys/aggr-[po6]"
AS_CALLBACK=/home/admin/active_standby/client1_callback.sh
```

Then you can just call the script with

`python2.7 active_standby.py --apic-password=${APICPASS}`

cornec81 replied:
i preferred the predictability albeit horrendously long cli to trusting envvars

Commit:
- made python3 friendly

Commit:
- added version >= 2.0.0 requirement for requests package

Commit:
- added PYTHON var that can be used to specify python version. default to python3.4

Commit:
updated active_standby.py:
- made python3 friendly.
- added logmsg() to use for timestamping output.
- added pc_attr() which returns aggregate interface info.
- changed callback script handling, to show error output.

Commit:
logrotate sample file

ghost commented:
regarding your comment from Apr 7 containing "sticky situation":
extra check has been added to only shut the standby, if the active is still up - reducing the possible race condition when running multiple scripts.
